packages/core/mesh/network-manager/src/transport/libdatachannel-transport.ts

```diff
@@ -0,0 +1,241 @@
+          }
+          this._channel = this._channelHandler(event.channel);
+        };
+      }
+
+      return peer;
+    })();
+  }
+
+  private _channelHandler(dataChannel: RTCDataChannel) {
+    dataChannel.onopen = () => {
+      log.info('dataChannel.onopen');
+      const duplex = new Duplex({
+        read: () => {},
+        write: (chunk, encoding, callback) => {
+          // todo wait to open
+
+          if (chunk.length > MAX_MESSAGE_SIZE) {
+            this.errors.raise(Error(`message too large: ${chunk.length} > ${MAX_MESSAGE_SIZE}`));
+          }
+          dataChannel.send(chunk);
+
+          if (this._channel.bufferedAmount > MAX_BUFFERED_AMOUNT) {
+            log.debug('bufferedAmount > MAX_BUFFERED_AMOUNT: PAUSING');
+            if (this._writeCallback !== null) {
+              log.error("consumer trying to write before we're ready for more data");
+            }
+            this._writeCallback = callback;
+          } else {
+            callback();
+          }
+        },
+      });
+
+      duplex.pipe(this.params.stream).pipe(duplex);
+      this._stream = duplex;
+      this._connected = true;
+      this.connected.emit();
+    };
+
+    dataChannel.onerror = async (err) => {
+      log.error('channel error', { err });
+      await this._disconnectStreams();
+      this._closed = true;
+      this.closed.emit();
+    };
+    dataChannel.onclose = async (err) => {
+      log.error('channel error', { err });
+      await this._disconnectStreams();
+      this._closed = true;
+      this.closed.emit();
+    };
+    dataChannel.onmessage = (event) => {
+      let data = event.data;
+      if (data instanceof ArrayBuffer) {
+        data = Buffer.from(data);
+      }
+      this._stream.push(data);
+    };
+
+    // TODO(mykola): Check if this is working correctly.
+    dataChannel.onbufferedamountlow = () => {
+      const cb = this._writeCallback;
+      this._writeCallback = null;
+      cb?.();
+    };
+
+    dataChannel.onclose = () => {
+      this._stream.destroy();
+      this._closed = true;
+      this.closed.emit();
+    };
+    return dataChannel;
+  }
+
+  signal(signal: Signal): void {
+    this._peer
+      .then(async (peer) => {
+        const data = signal.payload.data;
+        switch (data.type) {
+          case 'offer': {
+            if ((await this._peer).connectionState !== 'new') {
+              log.error('received offer but peer not in state new', { peer });
+              this.errors.raise(Error('invalid signalling state: received offer when peer is not in state new'));
+              break;
+            }
+            await peer.setRemoteDescription({ type: data.type, sdp: data.sdp });
+            const answer = await peer.createAnswer();
+            await peer.setLocalDescription(answer);
+            await this.params.sendSignal({ payload: { data: { type: answer.type, sdp: answer.sdp } } });
+            this._readyForCandidates.wake();
+            break;
```

Comment:
any error in this`offer` branch of switch-case statement is critical. Please wrap it in a try-catch statement and push those errors upstream with `errors.raise`